### PR TITLE
use point radius in extent definition

### LINKED ISSUE #2101: Slow viewer with many, big points close to each other.
## 🐛 Bug

If a points layer includes a lot of points whose size is big relatively to their relative distance, the viewer becomes slow and chockes up the system resources. ~This does not happen in 0.4.2~.

## To Reproduce

```python
import napari
import numpy as np
v = napari.Viewer(ndisplay=3)
# lots of overlapping points with coordinates between 0 and 1, and default size of 10
v.add_points(np.random.rand(10000, 3))
```

however, this works fine:
```python
import napari
import numpy as np
v = napari.Viewer(ndisplay=3)
# size = 0.01
v.add_points(np.random.rand(10000, 3), size=0.01)
```

## Expected behavior

Size does not matter :)

## Environment
```yaml
napari: 0.4.2
Platform: Linux-5.10.6-arch1-1-x86_64-with-glibc2.32
Python: 3.9.1 (default, Dec 13 2020, 11:55:53) [GCC 10.2.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.19.5
SciPy: 1.6.0
Dask: 2020.12.0
VisPy: 0.6.6

GL version: 4.6.0 NVIDIA 460.32.03
MAX_TEXTURE_SIZE: 16384

Plugins:
- mrcfile-reader: 0.1.2
- napari_plugin_engine: 0.1.8
- svg: 0.1.4
```

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR modifies the definition of the data extent in the `Points` layer to the axis aligned bounding box encompassing the displayed size of the points (i.e., it accounts for the size of the point). 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

@sofroniewn @brisvag could you check?

### CODE REVIEW

File: napari/layers/points/points.py
Comment (sofroniewn) [inline napari/layers/points/points.py]:
```suggestion
            maxs = np.max(self.data + self.size, axis=0)
            mins = np.min(self.data - self.size, axis=0)
```

Comment (sofroniewn) [reply napari/layers/points/points.py]: this would take the full size into account but might be slow

Comment (Czaki) [reply napari/layers/points/points.py]: done

File: napari/components/_tests/test_viewer_model.py
Comment (Czaki) [inline napari/components/_tests/test_viewer_model.py]: I'm not sure if I should fix this test in this way, or change somehow reset function?

Comment (sofroniewn) [reply napari/components/_tests/test_viewer_model.py]: Can this removed?

**Review by sofroniewn — APPROVED** [review]
I think this is fine. @kevinyamauchi @brisvag can you review this too? I'd be fine to merge if you agree

File: napari/components/_tests/test_viewer_model.py
Comment (Czaki) [reply napari/components/_tests/test_viewer_model.py]: if we even still have dims object in viewer? I think that this line was connected with the wrong reset button working when having only a points layer. 

I will merge the current master to this and then check.

Comment (sofroniewn) [reply napari/components/_tests/test_viewer_model.py]: thanks yeah - it seems unrelated to PR

**Comment by brisvag** [comment]
I'm not sure how much related, but #3444 talked and dealt with some of this. We should take a look at that before we go over the same stuff.

File: napari/components/_tests/test_viewer_model.py
Comment (Czaki) [reply napari/components/_tests/test_viewer_model.py]: it is totally related to PR. This is the fix of the test. The reason for this line is that on adding the first layer `viewer.dims.point` is set to the lower corner.

before these lines this condition is satisfied 
```python
assert viewer.dims.point == (-10, -10, -10, -9)
```

and `viewer.layers[labels_name]._data_raw`is empty as added points will be not visible.

File: napari/layers/points/points.py
Comment (kevinyamauchi) [inline napari/layers/points/points.py]: Should these be `self.size / 2`? I think the point size sets the diameter. Is that right, @brisvag ?

File: napari/components/_tests/test_viewer_model.py
Comment (sofroniewn) [reply napari/components/_tests/test_viewer_model.py]: > The reason for this line is that on adding the first layer viewer.dims.point is set to the lower corner.

interesting, i hadn't realized that. can you maybe add that as a comment. it seems easy to forget. thanks!!

Comment (Czaki) [reply napari/components/_tests/test_viewer_model.py]: Could You check it?

File: napari/layers/points/points.py
Comment (brisvag) [reply napari/layers/points/points.py]: Yup, size is the diameter!

Comment (Czaki) [reply napari/layers/points/points.py]: done

File: napari/layers/points/_tests/test_points.py
Comment (kevinyamauchi) [inline napari/layers/points/_tests/test_points.py]: I think it would be more clear if we explicitly set the point size.
```suggestion
    layer = Points(data, size=10)
```

File: napari/components/_tests/test_viewer_model.py
Comment (kevinyamauchi) [inline napari/components/_tests/test_viewer_model.py]: a few minor suggestions to clarify this comment

```suggestion
    # after adding a layer,  the dims point is set to lower `extent.world`.
    # thus after adding the points data, the dims point is (-5.0, -5.0, -5.0, -4.0)
    # Therefore, we need to reset position to see data from slice `[0, 0, ...]`
    # as required by the tests below
```

**Review by kevinyamauchi — COMMENTED** [review]
Thanks for the updates and comments, @Czaki !

After seeing the required fix in the test, I am a little hesitant about changing the definition of the extents of the points layer to include the point sizes. If I understand correctly, if a user adds a single point, it would not be visible in the viewer because the `dims.point` would be set at the negative extent of the point (e.g., `(-5, -5, -5)` for a point at `(0, 0, 0)` with size 10). Is this correct?

I guess this goes back to the discussion in #3444 pointed out by @brisvag . Is layer extent about slicing or is it about the display volume? Thoughts?

I am sorry for the back and forth @Czaki .

File: napari/components/_tests/test_viewer_model.py
Comment (Czaki) [reply napari/components/_tests/test_viewer_model.py]: Thanks

**Comment by Czaki** [comment]
> if a user adds a single point, it would not be visible in the viewer because the `dims.point` would be set at the negative extent of the point

Yes. The same problem is when you load the tiff file with empty bottom layer (sample file attached).
[A_component17.zip](https://github.com/napari/napari/files/7434585/A_component17.zip)

In PartSeg I set point to middle of loaded image, instead of on bottom. I prefer such solution to set starting position.

**Comment by Czaki** [comment]
macOS fail looks related to the problem with global action manager nad problem of partially destroyed qt objects (#3437).

**Comment by Czaki** [comment]
@brisvag could you check if thsi solve #2101. Until few days ago this PR was marked as draft so maybe this not fully solve this issue.

**Comment by brisvag** [comment]
@Czaki this does not solve that problem, it seems, but it does make the initialized zoom a bit more sensible (now this code shows the whole point layer rather than being zoomed in so much that everything is white).

```python
import napari
import numpy as np
v = napari.Viewer(ndisplay=3)
# lots of overlapping points with coordinates between 0 and 1, and default size of 10
v.add_points(np.random.rand(10000, 3))
```

This does have a slight benefit to performance byt not being so zoomed in, but I think that the only way to actually solve that problem is to figure out why vispy gets such a big performance hit in the first place. It seems that the performance drop is proportional to: `(size * number_of_points) / distances`

**Comment by Czaki** [comment]
@sofroniewn @kevinyamauchi could you unpin issue #2101 from this PR? 

I think that the partially slow down may be connected with the calculation of which point should be rendered in front. Does Mabe switch to render points as the ball could solve this?

**Comment by kevinyamauchi** [comment]
> @sofroniewn @kevinyamauchi could you unpin issue #2101 from this PR?

I updated the PR description so it no longer references #2101. I believe this should "unlink it" (i.e., when we merge this,  #2101 won't be automatically closed). Is that what you had in mind?

**Comment by kevinyamauchi** [comment]
> In PartSeg I set point to middle of loaded image, instead of on bottom. I prefer such solution to set starting position.

Hmm, I think that makes sense to me as well. I've made an issue here: https://github.com/napari/napari/issues/3559. I couldn't find the part of the code where the `Dims.point` gets updated by adding a new layer. If you know where that is, would you mind dropping a link in that issue?

**Comment by brisvag** [comment]
> I think that the partially slow down may be connected with the calculation of which point should be rendered in front. Does Mabe switch to render points as the ball could solve this?

This woulkd be strange; this kind of calculation happens in the shader thanks to the depth buffer, and as far as I know comparing two numbers really close to each other or two number very far from each other should not impact performance?

**Comment by sofroniewn** [comment]
Given that this doesn't solve #2101 do we still want to merge this @kevinyamauchi @brisvag? I see that #3559 has now been made. Should that be addressed first? And then we merge this one? Or do we just close this one?

File: napari/layers/points/points.py
Comment (grlee77) [inline napari/layers/points/points.py:602]: Nice, I was tempted to make this change when looking #3444, but didn't want to introduce other new things in that PR. This seems consistent with incorporating the full pixel widths in the extent for image layers.

File: napari/components/_tests/test_viewer_model.py
Comment (grlee77) [inline napari/components/_tests/test_viewer_model.py:128]: I think it is a little less clear what `dims.range` should be for points than for image-based layers (see for example discussion here: https://github.com/napari/napari/pull/3444#discussion_r720607644.

As a concrete example, if I check out this PR branch and run the following script that generates a single 3D point at the origin, nothing is visible when it first opens, but if I scroll the slider to 5th position to see the point.

```Python
import numpy as np

import napari
from napari import Viewer

with napari.gui_qt():
    viewer = Viewer()

    one_pixel = np.ones((1, 1, 1))
    point = np.zeros((1, 3))
    viewer.add_points(point)

```

I suppose, if the change proposed in #3559 (set initial point to the centroid) was also made it would make the point visible upon first opening the viewer, though.

**Comment by grlee77** [comment]
> Given that this doesn't solve #2101 do we still want to merge this @kevinyamauchi @brisvag? I see that #3559 has now been made. Should that be addressed first? And then we merge this one? Or do we just close this one?

I think we may want to try #3559 first and then update any tests as needed in this PR.

**Comment by kevinyamauchi** [comment]
I agree with @grlee77 that we should first update the default slice to the midpoint of the world extents (#3559) and then circle back to this PR. I think with #3559, we would no longer need to manually set the the dims point in the tests as is done in this PR (https://github.com/napari/napari/pull/2104#discussion_r557310626)

Thank you for your patience @Czaki as we figure out the best way forward.

**Comment by sofroniewn** [comment]
> I think we may want to try #3559 first and then update any tests as needed in this PR.

Agreed - @grlee77 can you give that a go in a new PR, thanks!

**Comment by sofroniewn** [comment]
now that #3718 is merged @Czaki do you want to fix there merge conflict here. Will this then be ready for review? Maybe @grlee77 can review then. Thanks all!!

**Comment by grlee77** [comment]
@Czaki, if you are busy, I can also try to resume from here in a new PR. Just let me know

**Comment by Czaki** [comment]
I live in another timezone, and not spend all evenings on coding.

**Review by sofroniewn — APPROVED** [review]
LGTM! Thanks for fix @Czaki, thanks for offering to help @grlee77. Will let some others have a chance to look too - @kevinyamauchi @brisvag. If all good then lets merge.

**Review by grlee77 — APPROVED** [review]
Looks great, thanks @Czaki.

(My earlier offer to help was not meant as impatience or to rush you, just an indication of willingness to help out)

**Review by brisvag — APPROVED** [review]
Code looks good to me, and everything seems to work.

I have to say I'm still not sold on point size being used for the extent; If I pick a point on the edge of a volume, it would feel strange to move the slider PAST the image extent, ointo a world where only a part of the point exists. This has similar issues to #3734: points do not always have a *physical* extent themselves, and size can be used to convey information other than physical size.

Having said this, if everyone is on board with this, I don't feel strongly enough about this to block it!

**Comment by brisvag** [comment]
Here's an example of what I mean:

```python
import napari
import numpy as np

img = np.random.rand(100, 100, 100)
pts = np.array([[0, 0, 0]])
v = napari.Viewer()
v.add_points(pts)
v.add_image(img)
```


https://user-images.githubusercontent.com/23482191/145555749-fe9e869e-70b9-4f99-97b0-a2b7572a4028.mp4

**Comment by Czaki** [comment]
@brisvag I do not know why it was removed by @kevinyamauchi  but this pr starts as a try to fix  #2101, which is an issue open by you. I see the problem pointed by you, but do not see how to solve  #2101 without introducing this.

**Comment by brisvag** [comment]
@Czaki  we already concluded in the comments above that this does not really solve that issue:

> @Czaki this does not solve that problem, it seems, but it does make the initialized zoom a bit more sensible (now this code shows the whole point layer rather than being zoomed in so much that everything is white).

I that issue is the *only* justification for this change, then I'm against merging this as is. Maybe we could add a flag somewhere so users can activate it only when point sizes have a physical extent.

An alternative to solve #2101 is to limit the zoom of the camera, no mattter how small the extent is, but this would not work with widldly different point sizes from the default.

**Comment by sofroniewn** [comment]
What does @kevinyamauchi, @jni or @alisterburt think about this https://github.com/napari/napari/pull/2104#pullrequestreview-828689479? It would be good to get another opinion here before merge

**Comment by alisterburt** [comment]
Ooft, this is difficult - @kevinyamauchi and I came across the same issue when doing point click and drag and our solution was to use the increased extents only where necessary. This avoids side effects like @brisvag noted but isn't easily reusable for other purposes the way we did it...

It feels a bit unintuitive to me and I don't think I would like to see this merged as is, it's interesting to think about though so thank you @Czaki !

Maybe a more concrete solution is to have a model for the extents which includes 
- data extents in data coordinates
- data extents in world coordinates
- depiction extents in the displayed world coordinates (2d or 3d, including extension for half pixels/point radii)

These could be named extents.data, extends.world and extents.depicted_world

Thoughts?

**Comment by sofroniewn** [comment]
> depiction extents in the displayed world coordinates (2d or 3d, including extension for half pixels/point radii)

We could add this, but then we'd have to decide when to use this vs other ones. I'm not sure. @jni?

**Comment by brisvag** [comment]
Yeah I agree with @alisterburt ! Not a huge fan of "depicted" though, it's not clear what it refers to. Just now this made me think: that method that you and @kevinyamauchi added in your 3D interactivity PR called `_display_bounding_box`, does it do basically this?

I'd go with something more explicit such as `extent_with_sizes`. Still, what would we do with that? Pass both to the viewer, and use the `extent_with_sizes` as a lower bound for setting the zoom?

**Review by brisvag — CHANGES_REQUESTED** [review]
Gonna mark this as un-approved while we decide what to do :)

**Comment by sofroniewn** [comment]
@Czaki can you provide some more detail here on what you're trying to achieve - is this more about the 2D visual display when you reset the zoom? the 2D drawing of the bounding box? or the nD scrolling case - like @brisvag pointed out here https://github.com/napari/napari/pull/2104#issuecomment-990801175. We could improve the nD case by only enabling this mode if `n_dimensional` is True. But that wouldn't address the 2D only case unless one turned it on.

**Comment by Czaki** [comment]
@sofroniewn as I pointed [here](https://github.com/napari/napari/pull/2104#issuecomment-990915098) this PR is trying to fix of #2101. Nothing else.   
So the clue is to be sure that the view area after reset/loading multiple close points is enough big to see whole points. Maybe yes, maybe best will somehow apply only when calculating the bounding box for the default view. But it may introduce another level of complication to architecture.

**Comment by sofroniewn** [comment]
> @sofroniewn as I pointed here this PR is trying to fix of #2101. Nothing else.
So the clue is to be sure that the view area after reset/loading multiple close points is enough big to see whole points. Maybe yes, maybe best will somehow apply only when calculating the bounding box for the default view. But it may introduce another level of complication to architecture.

Ok thanks - maybe @jni can you try and reproduce #2101 and then weigh in on the discussion above. I'd like us to hopefully be able to resolve this so that it doesn't drag out indefinitely

**Comment by sofroniewn** [comment]
It looks like we will not be merging this PR as is, and so I'd like to close it in favour of continued discussion in #3848 and #2101. Please let me know if I shouldn't do this, otherwise if i hear nothing, in a few days I will close it

**Comment by Czaki** [comment]
Close.

**Pull request closed without merging**
